crates/multi-transaction-payment/src/lib.rs

```diff
@@ -0,0 +1,169 @@
+mod tests;
+
+type CallOf<T> = <T as Config>::RuntimeCall;
+type SubstrateDefaultPayment<T> =
+    pallet_transaction_payment::CurrencyAdapter<<T as Config>::Currency, <T as Config>::OnUnbalanced>;
+
+#[frame_support::pallet]
+pub mod pallet {
+    use super::*;
+    use frame_support::{
+        dispatch::{DispatchInfo, GetDispatchInfo, PostDispatchInfo},
+        pallet_prelude::*,
+    };
+    use frame_system::pallet_prelude::*;
+    use sp_runtime::traits::Dispatchable;
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config<RuntimeCall = CallOf<Self>> + currency::Config<Balance = u128> {
+        /// The aggregated call type.
+        type RuntimeCall: Parameter
+            // + Dispatchable<RuntimeOrigin = Self::RuntimeOrigin, PostInfo = PostDispatchInfo, Info = DispatchInfo>
+            + Dispatchable<RuntimeOrigin = Self::RuntimeOrigin, PostInfo = PostDispatchInfo, Info = DispatchInfo>
+            + GetDispatchInfo
+            + IsSubType<Call<Self>>;
+
+        /// dex api
+        type Dex: ExportDexGeneral<Self::AccountId, Self::CurrencyId>;
+
+        /// The currency trait.
+        type Currency: Currency<Self::AccountId, Balance = u128>;
+
+        /// What to do with fees. Typically this is a transfer to the block author or treasury
+        type OnUnbalanced: OnUnbalanced<<Self::Currency as Currency<Self::AccountId>>::NegativeImbalance>;
+
+        /// weights of dex operations
+        type DexWeightInfo: dex_general::WeightInfo;
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {}
+
+    #[pallet::pallet]
+    pub struct Pallet<T>(_);
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        #[pallet::call_index(0)]
+        #[pallet::weight({
+			let dispatch_info = call.get_dispatch_info();
+			((T::DexWeightInfo::swap_assets_for_exact_assets(_path.len() as u32)).saturating_add(dispatch_info.weight), dispatch_info.class,)
+		})]
+        #[frame_support::transactional]
+        pub fn with_fee_swap_path(
+            origin: OriginFor<T>,
+            _path: Vec<T::CurrencyId>,
+            _amount_in_max: <T as currency::Config>::Balance,
+            call: Box<CallOf<T>>,
+        ) -> DispatchResultWithPostInfo {
+            // Note: no swaps are being done here - they have already been done in the `pre_dispatch`
+            // of the SignedExtension, which calls the `OnChargeTransaction::withdraw_fee` implemented below.
+
+            let mut ret = call.dispatch(origin);
+            // modify any returned weight overrides. Note that we pass-through the `Pays` unmodified.
+            let modify_weight = |x: &mut PostDispatchInfo| {
+                x.actual_weight = x
+                    .actual_weight
+                    .map(|x| x.saturating_add(T::DexWeightInfo::swap_assets_for_exact_assets(_path.len() as u32)));
+            };
+            match ret {
+                Ok(ref mut info) => {
+                    // if a weight override is returned, add the cost of the swap
```

Comment:
If this code compiled it means that the return value of the dispatch is always a `PostDispatchInfo`, right? In that case I'd remove the `if` in this comments to avoid confusion
```suggestion
                    // a weight override is returned, add the cost of the swap
```

Reply:
The return value is always `PostDispatchInfo` indeed, but `PostDispatchInfo.actual_weight` is an Option - it may or may not return a weight override

Reply:
Then maybe move this comment inside the `modify_weight` closure?